consumer_monitor.go

```diff
@@ -0,0 +1,280 @@
+	if offsetFetchInterval <= 0 || offsetFetchInterval > 10*time.Minute {
+		offsetFetchInterval = 3 * time.Minute
+	}
+	maxFailureCount := config.OffsetFetchMaxFailureCount
+	if maxFailureCount <= 0 || maxFailureCount > 10 {
+		maxFailureCount = 5
+	}
+
+	return &consumerMonitor{
+		consumerGroup: config.ConsumerGroup,
+		scheduler: fetcherScheduler{
+			standardInterval:  offsetFetchInterval,
+			shortenedInterval: offsetFetchInterval / 3,
+			maxFailureCount:   maxFailureCount,
+		},
+		consumerOffsetFetcher: consumerFetcher,
+		topicOffsetFetcher:    topicFetcher,
+		subscriptions:         map[string][]int32{},
+		topicsLock:            &sync.RWMutex{},
+		topics:                topics,
+		logger:                logger,
+	}
+}
+
+func (m *consumerMonitor) run(ctx context.Context, subscriptionEvents chan *subscriptionEvent) {
+	m.scheduler.ticker = time.NewTicker(m.scheduler.standardInterval)
+	defer m.scheduler.ticker.Stop()
+
+	log := m.logger.WithField("process", "ConsumerMonitor")
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Terminate the fetchers and exit.
+			// Note that both fetchers share the same connection so calling Close() once is enough.
+			if err := m.consumerOffsetFetcher.Close(); err != nil {
+				log.WithError(err).Error("Failed to close offset fetcher connections")
+			}
+
+			log.Info("Terminating consumer monitor...")
+			return
+
+		case event := <-subscriptionEvents:
+			log.
+				WithField("subscribed", event.subscribed).
+				WithField("topic", event.topic).
+				WithField("partition", event.partition).
+				Info("Subscription event received")
+
+			m.updateSubscriptions(event)
+
+		case <-m.scheduler.ticker.C:
+			if len(m.subscriptions) == 0 {
+				log.Warn("Consumer is not currently subscribed for any topics")
+				continue
+			}
+
+			offsets, err := m.fetchOffsets()
+			if err != nil {
+				log.WithError(err).Error("Failed to fetch consumer offsets")
+
+				m.scheduler.failureCount++
+				if m.scheduler.failureCount >= m.scheduler.maxFailureCount {
+					log.Errorf("Fetching offsets failed %d times in a row. Consumer status data is stale.",
+						m.scheduler.failureCount)
+
+					m.clearConsumerStatus()
+				}
+
+				m.scheduler.ticker.Reset(m.scheduler.shortenedInterval)
+				continue
+			}
+
+			log.WithField("offsets", offsets).Debug("Offsets fetched successfully")
+
+			m.updateConsumerStatus(offsets)
+
+			if m.scheduler.failureCount != 0 {
+				// Revert the scheduling change.
+				m.scheduler.failureCount = 0
+				m.scheduler.ticker.Reset(m.scheduler.standardInterval)
+			}
+		}
+	}
+}
+
+func (m *consumerMonitor) updateSubscriptions(event *subscriptionEvent) {
+	if event.subscribed {
+		m.subscriptions[event.topic] = append(m.subscriptions[event.topic], event.partition)
+		return
+	}
+
+	partitions, ok := m.subscriptions[event.topic]
+	if !ok {
+		return
+	}
+
+	for i, p := range partitions {
+		if p == event.partition {
+			partitions = append(partitions[:i], partitions[i+1:]...)
```

Comment:
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**potential issue:** ‏With `partitions[:i]` we are reslicing here, [using `append` with different slice than its result could lead to subtle bugs](https://medium.com/@Jarema./golang-slice-append-gotcha-e9020ff37374).

Reply:
That's a great example of a problem I've never thought possible. In our case though we should be fine since re-slicing is a convenient option for removing an element using its index.